NullPointerException when using custom Region or AWS_GLOBAL in S3Client

### Describe the bug

When using a custom region (example, Region.of("custom")  or Region.AWS_GLOBAL a NullPointerException is thrown when resolving a S3 Bucket endpoint. This impacts the use case of using the SDK to communicate with other S3 services where Region is ignored. Since Region is required in the S3ClientBuilder someone may want to do something like

```
S3Client.builder().region(Region.of("ignored")...
```

which was supported in previous versions of the SDK.

interesting enough this issue also exists if you are using Region.AWS_GLOBAL

### Expected behavior

The expected behavior is that the NullPointerException is not thrown and requests are sucessful using custom regions.

### Current behavior

NullPointerException is thrown

```
Caused by: java.lang.NullPointerException: null
	at software.amazon.awssdk.services.s3.internal.endpoints.S3BucketEndpointResolver.resolveEndpoint(S3BucketEndpointResolver.java:72) ~[s3-2.17.83.jar:na]
	at software.amazon.awssdk.services.s3.internal.endpoints.S3BucketEndpointResolver.applyEndpointConfiguration(S3BucketEndpointResolver.java:50) ~[s3-2.17.83.jar:na]
	at software.amazon.awssdk.services.s3.internal.handlers.EndpointAddressInterceptor.modifyHttpRequest(EndpointAddressInterceptor.java:68) ~[s3-2.17.83.jar:na]
	at software.amazon.awssdk.core.interceptor.ExecutionInterceptorChain.modifyHttpRequestAndHttpContent(ExecutionInterceptorChain.java:90) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.core.internal.handler.BaseClientHandler.runModifyHttpRequestAndHttpContentInterceptors(BaseClientHandler.java:139) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.core.internal.handler.BaseClientHandler.finalizeSdkHttpFullRequest(BaseClientHandler.java:77) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.doExecute(BaseSyncClientHandler.java:151) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.lambda$execute$1(BaseSyncClientHandler.java:82) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.measureApiCallSuccess(BaseSyncClientHandler.java:175) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.core.internal.handler.BaseSyncClientHandler.execute(BaseSyncClientHandler.java:76) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.core.client.handler.SdkSyncClientHandler.execute(SdkSyncClientHandler.java:45) ~[sdk-core-2.17.83.jar:na]
	at software.amazon.awssdk.awscore.client.handler.AwsSyncClientHandler.execute(AwsSyncClientHandler.java:56) ~[aws-core-2.17.83.jar:na]
	at software.amazon.awssdk.services.s3.DefaultS3Client.headBucket(DefaultS3Client.java:4901) ~[s3-2.17.83.jar:na]
```

### Steps to Reproduce

Use a custom region when building S3Client and send a request.

```
        final S3Client s3Client = S3Client.builder().
                .region(Region.of("custom").build();
        s3Client.headBucket(HeadBucketRequest.builder().bucket("myBucket").build());
```

### Possible Solution

This seems to be the commit that caused the issue.

https://github.com/aws/aws-sdk-java-v2/commit/4656690aed8effe36fcf31cec695778992ee8228#diff-e54f09e0d313a4de0a7e7590ad1dcd7a7d36a943aece2f74e2c41592f74dd081


A possible solution is to change this line
 `String dnsSuffixWithoutTagConsideration = RegionMetadata.of(context.region()).domain();`

so that domain() is not called if the region meta data is null

### Context

_No response_

### AWS Java SDK version used

2.17.83

### JDK version used

11.0.13

### Operating System and version

Windows 7

--- a/.changes/next-release/bugfix-AmazonS3-3b61c48.json
+++ b/.changes/next-release/bugfix-AmazonS3-3b61c48.json
@@ -1,0 +1,6 @@
+{
+    "category": "Amazon S3", 
+    "contributor": "", 
+    "type": "bugfix", 
+    "description": "Fixed an issue that resulted in a NullPointerException when an invalid or global region was used on an S3 client. Fixes [#2885](https://github.com/aws/aws-sdk-java-v2/issues/2885)."
+}

--- a/services/s3/src/it/java/software/amazon/awssdk/services/s3/ListBucketsIntegrationTest.java
+++ b/services/s3/src/it/java/software/amazon/awssdk/services/s3/ListBucketsIntegrationTest.java
@@ -1,0 +1,28 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3;
+
+import org.junit.Test;
+import software.amazon.awssdk.regions.Region;
+
+public class ListBucketsIntegrationTest extends S3IntegrationTestBase {
+    @Test
+    public void listBuckets_InGlobal_DoesNotThrowException() {
+        try (S3Client s3 = s3ClientBuilder().region(Region.AWS_GLOBAL).build()) {
+            s3.listBuckets();
+        }
+    }
+}

--- a/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/endpoints/S3BucketEndpointResolver.java
+++ b/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/endpoints/S3BucketEndpointResolver.java
@@ -26,6 +26,7 @@ import static software.amazon.awssdk.utils.FunctionalUtils.invokeSafely;
 import java.net.URI;
 import software.amazon.awssdk.annotations.SdkInternalApi;
 import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.regions.PartitionMetadata;
 import software.amazon.awssdk.regions.RegionMetadata;
 import software.amazon.awssdk.services.s3.S3Configuration;
 import software.amazon.awssdk.services.s3.internal.BucketUtils;
@@ -69,7 +70,10 @@ public final class S3BucketEndpointResolver implements S3EndpointResolver {
     private static URI resolveEndpoint(S3EndpointResolverContext context) {
         SdkHttpRequest request = context.request();
         String protocol = request.protocol();
-        String dnsSuffixWithoutTagConsideration = RegionMetadata.of(context.region()).domain();
+        RegionMetadata regionMetadata = RegionMetadata.of(context.region());
+        String dnsSuffixWithoutTagConsideration = regionMetadata != null ?
+                                                  regionMetadata.domain() :
+                                                  PartitionMetadata.of(context.region()).dnsSuffix();
         S3Configuration serviceConfiguration = context.serviceConfiguration();
 
         boolean useAccelerate = isAccelerateEnabled(serviceConfiguration) && isAccelerateSupported(context.originalRequest());

--- a/services/s3/src/test/java/software/amazon/awssdk/services/s3/InvalidRegionTest.java
+++ b/services/s3/src/test/java/software/amazon/awssdk/services/s3/InvalidRegionTest.java
@@ -22,6 +22,7 @@ import org.junit.Test;
 import software.amazon.awssdk.auth.credentials.AnonymousCredentialsProvider;
 import software.amazon.awssdk.core.exception.SdkClientException;
 import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.model.HeadBucketRequest;
 import software.amazon.awssdk.services.s3.presigner.S3Presigner;
 
 public class InvalidRegionTest {
@@ -45,6 +46,17 @@ public class InvalidRegionTest {
             .hasMessageContaining("US_WEST_2")
             .hasMessageContaining("region")
             .hasMessageContaining("us-west-2");
+    }
+
+    @Test
+    public void nonExistentRegionGivesHelpfulMessage() {
+        S3Client s3Client = S3Client.builder()
+                                    .region(Region.of("does-not-exist"))
+                                    .credentialsProvider(AnonymousCredentialsProvider.create())
+                                    .build();
+        assertThatThrownBy(() -> s3Client.headBucket(HeadBucketRequest.builder().bucket("myBucket").build()))
+            .isInstanceOf(SdkClientException.class)
+            .hasMessageContaining("UnknownHostException");
     }
 
     @Test

